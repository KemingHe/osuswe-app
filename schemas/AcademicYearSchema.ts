import { z } from 'zod';

export const ACADEMIC_YEAR_OPTIONS: string[] = [
  'High School',
  '1st Year (Freshman)',
  '2nd Year (Sophomore)',
  '3rd Year (Junior)',
  '4th Year (Senior)',
  '5th+ Year (Super Senior)',
];
const academicYearErrorMessage: string = 'Invalid academic year';
const isValidAcademicYear = (value: string): boolean =>
  ACADEMIC_YEAR_OPTIONS.includes(value);

export const AcademicYearSchema = z
  .string({ message: academicYearErrorMessage })
  .refine(isValidAcademicYear, { message: academicYearErrorMessage });

export type AcademicYear = z.infer<typeof AcademicYearSchema>;
