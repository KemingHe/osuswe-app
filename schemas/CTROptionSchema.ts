import { z } from 'zod';

export const CTR_OPTIONS: string[] = [
  'I am a regional campus student',
  'I am a transfer student',
  'I am a commuter student',
  'I am a Columbus State Community College student',
  'I am none of the above',
];
const ctrErrorMessage: string = 'Invalid CTR selection';
const isValidCTROption = (value: string): boolean =>
  CTR_OPTIONS.includes(value);

export const CTROptionSchema = z
  .string({ message: ctrErrorMessage })
  .refine(isValidCTROption, { message: ctrErrorMessage });

export type CTROption = z.infer<typeof CTROptionSchema>;
