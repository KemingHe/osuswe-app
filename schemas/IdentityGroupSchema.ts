import { z } from 'zod';

export const IDENTITY_GROUP_OPTIONS: string[] = [
  'African-American',
  'Asian Connection',
  'HeForSWE (male allies of women in engineering)',
  'Latinx',
  'LGBTQ+',
  'Native-American/Indigenous Peoples',
  'Military and Veterans',
  'Disability Inclusion',
  'I do not identify with any of the above groups',
];
const identityGroupErrorMessage: string = 'Invalid identity group selection';
const isValidIdentityGroup = (value: string): boolean =>
  IDENTITY_GROUP_OPTIONS.includes(value);

export const IdentityGroupSchema = z
  .string({ message: identityGroupErrorMessage })
  .refine(isValidIdentityGroup, { message: identityGroupErrorMessage });

export type IdentityGroup = z.infer<typeof IdentityGroupSchema>;
