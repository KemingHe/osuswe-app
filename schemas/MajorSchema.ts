import { z } from 'zod';

export const MAJOR_OPTIONS: string[] = [
  'Aerospace Engineering',
  'Aviation',
  'Biomedical Engineering',
  'Chemical Engineering',
  'Civil Engineering',
  'Computer Science and Engineering',
  'Electrical and Computer Engineering',
  'Engineering Physics',
  'Engineering Technology',
  'Environmental Engineering',
  'Food, Agricultural, and Biological Engineering',
  'Industrial and Systems Engineering',
  'Materials Science and Engineering',
  'Mechanical Engineering',
  'Welding Engineering',
  'Undecided',
  'Architecture',
  'Other',
];
const majorErrorMessage: string = 'Invalid major selection';
const isValidMajor = (value: string): boolean => MAJOR_OPTIONS.includes(value);

export const MajorSchema = z
  .string({ message: majorErrorMessage })
  .refine(isValidMajor, { message: majorErrorMessage });

export type Major = z.infer<typeof MajorSchema>;
