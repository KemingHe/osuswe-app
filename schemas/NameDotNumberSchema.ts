import { NAME_DOT_NUM_PATTERN } from '@keminghe/osu';
import { z } from 'zod';

const invalidErrorMessage: string = 'Enter a valid name.#';

export const NameDotNumberSchema = z
  .string({ message: invalidErrorMessage })
  .regex(NAME_DOT_NUM_PATTERN, { message: invalidErrorMessage });

export type NameDotNumber = z.infer<typeof NameDotNumberSchema>;
