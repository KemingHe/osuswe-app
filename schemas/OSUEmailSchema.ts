import { OSU_DOT_EDU_EMAIL_PATTERN } from '@keminghe/osu';
import { z } from 'zod';

const invalidErrorMessage: string = 'Enter a valid @osu.edu email';

export const OSUEmailSchema = z
  .string({ message: invalidErrorMessage })
  .regex(OSU_DOT_EDU_EMAIL_PATTERN, { message: invalidErrorMessage });

export type OSUEmail = z.infer<typeof OSUEmailSchema>;
