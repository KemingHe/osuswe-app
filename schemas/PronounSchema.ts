import { z } from 'zod';

export const PRONOUN_OPTIONS: string[] = [
  'he/him',
  'she/her',
  'they/them',
  'ze/zir',
  'other',
  'prefer not to say',
];
const pronounErrorMessage: string = 'Invalid pronoun';
const isValidPronoun = (value: string): boolean =>
  PRONOUN_OPTIONS.includes(value);

export const PronounSchema = z
  .string({ message: pronounErrorMessage })
  .refine(isValidPronoun, { message: pronounErrorMessage });

export type Pronoun = z.infer<typeof PronounSchema>;
