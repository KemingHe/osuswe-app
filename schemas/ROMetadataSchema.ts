import { z } from 'zod';

import { UserRoleSchema } from '@/schemas/UserRoleSchema';

const invalidErrorMessage: string = 'Invalid client read-only metadata';

export const ROMetadataSchema = z
  .object(
    {
      role: UserRoleSchema,
    },
    { message: invalidErrorMessage },
  )
  .strict({ message: invalidErrorMessage });

export type ROMetadata = z.infer<typeof ROMetadataSchema>;
