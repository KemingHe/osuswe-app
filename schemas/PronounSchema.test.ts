import { PRONOUN_OPTIONS, PronounSchema } from '@/schemas/PronounSchema';

describe('PRONOUN_OPTIONS', () => {
  it('matches the latest snapshot', () => {
    expect(PRONOUN_OPTIONS).toMatchSnapshot();
  });
});

describe('PronounSchema', () => {
  it('parses valid pronouns', () => {
    expect(PronounSchema.parse('he/him')).toBe('he/him');
    expect(PronounSchema.parse('she/her')).toBe('she/her');
    expect(PronounSchema.parse('they/them')).toBe('they/them');
    expect(PronounSchema.parse('ze/zir')).toBe('ze/zir');
    expect(PronounSchema.parse('other')).toBe('other');
    expect(PronounSchema.parse('prefer not to say')).toBe('prefer not to say');
  });

  it('throws an error for invalid pronouns', () => {
    expect(() => PronounSchema.parse('invalid')).toThrow();
    expect(() => PronounSchema.parse('')).toThrow();
  });

  it('throws an error for invalid types', () => {
    expect(() => PronounSchema.parse(123)).toThrow();
    expect(() => PronounSchema.parse({})).toThrow();
    expect(() => PronounSchema.parse([])).toThrow();
    expect(() => PronounSchema.parse(null)).toThrow();
    expect(() => PronounSchema.parse(undefined)).toThrow();
  });

  it('throws the correct error message for invalid string', () => {
    const result = PronounSchema.safeParse('invalid');
    expect(result.success).toBe(false);
    expect(result.error?.issues[0].message).toBe('Invalid pronoun');
  });

  it('throws the correct error message for valid type', () => {
    const result = PronounSchema.safeParse(123);
    expect(result.success).toBe(false);
    expect(result.error?.issues[0].message).toBe('Invalid pronoun');
  });
});
