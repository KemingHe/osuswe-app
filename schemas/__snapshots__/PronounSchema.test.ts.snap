// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PRONOUN_OPTIONS > matches the latest snapshot 1`] = `
[
  "he/him",
  "she/her",
  "they/them",
  "ze/zir",
  "other",
  "prefer not to say",
]
`;
